File: FE/src/types/application.ts
import type { Todo } from "./todo";
import type { DocumentItem } from "./document";

export const APPLICATION_STATUSES = [
  "작성 중",
  "결과 대기",
  "필기 전형",
  "면접 전형",
  "최종 결과",
] as const;

export const APPLICATION_FINAL_RESULTS = ["합격", "불합격", "포기"] as const;

export type ApplicationStatus = (typeof APPLICATION_STATUSES)[number];
export type ApplicationFinalResult =
  | (typeof APPLICATION_FINAL_RESULTS)[number]
  | null;

export type Application = {
  id: number;
  noticeId?: number | null;
  company: string;
  jobTitle: string;
  position: string;
  industry: string;
  applyDate?: string;
  interviewDate?: string;
  deadlineDate?: string;
  applyEventId?: string | null;
  interviewEventId?: string | null;
  deadlineEventId?: string | null;
  createdAt?: string;
  updatedAt?: string;
  status: ApplicationStatus;
  nextStep?: string;
  dday?: string | number;
  recentUpdated?: string;
  employmentType?: string;
  employType?: string;
  careerType?: string;
  jobType?: string;
  finalResult?: ApplicationFinalResult;
  sourceUrl?: string;
  url?: string;
  submitted?: boolean;
  checklistInComplete?: boolean;
  important?: boolean;
  documents?: DocumentItem[];
  memo?: string;
  todos?: Todo[];
};

export const COLUMN_OPTIONS = [
  { key: "position", label: "직무", default: true },
  { key: "employmentType", label: "고용형태", default: true },
  { key: "status", label: "현재 상태", default: true },
  { key: "deadlineDate", label: "마감일", default: true },
  { key: "dday", label: "D-day", default: true },
  { key: "checklistInComplete", label: "일정/할 일", default: true },
  { key: "industry", label: "산업", default: false },
  { key: "recentUpdated", label: "최근 수정일", default: true },
  { key: "createdAt", label: "등록일", default: true },
];

export const DEFAULT_COLUMNS = COLUMN_OPTIONS.filter(
  (column) => column.default,
).map((column) => column.key);

export type RegistrationTab = "URL" | "PDF" | "IMAGE" | "MANUAL";

File: BE/src/main/java/back/pickd/application/entity/Application.java
package back.pickd.application.entity;

import back.pickd.application.enums.ApplicationFinalResult;
import back.pickd.application.enums.ApplicationStatus;
import back.pickd.coverletter.entity.CoverLetterItem;
import back.pickd.document.entity.Document;
import back.pickd.notice.entity.Notice;
import back.pickd.user.entity.User;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "applications")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class Application {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "notice_id")
    private Notice notice;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ApplicationStatus status;

    @Enumerated(EnumType.STRING)
    @Column(length = 20)
    private ApplicationFinalResult finalResult;

    private String company;
    private String jobTitle;
    private String position;
    private String industry;
    private String memo;

    private LocalDateTime applyDate;
    private LocalDateTime interviewDate;
    private LocalDateTime deadlineDate;

    private String applyEventId;
    private String interviewEventId;
    private String deadlineEventId;

    @Column(nullable = false)
    @Builder.Default
    private boolean important = false;

    @Column(nullable = false)
    @Builder.Default
    private boolean manualRegistration = false;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Builder.Default
    @JsonManagedReference
    @OneToMany(mappedBy = "application", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Todo> todos = new ArrayList<>();

    @OneToMany(mappedBy = "application", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<Document> documents = new ArrayList<>();

    @OneToMany(mappedBy = "application", cascade = CascadeType.ALL, orphanRemoval = true)
    @Builder.Default
    private List<CoverLetterItem> coverLetterItems = new ArrayList<>();

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = this.createdAt;
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }

    // ── 도메인 메서드 ──────────────────────────────────────────────────────────

    public void update(String company, String jobTitle, String position, String industry,
                       ApplicationStatus status, ApplicationFinalResult finalResult,
                       boolean important, boolean manualRegistration, String memo,
                       LocalDateTime applyDate, LocalDateTime interviewDate, LocalDateTime deadlineDate) {
        this.company = company;
        this.jobTitle = jobTitle;
        this.position = position;
        this.industry = industry;
        this.status = status;
        this.finalResult = status == ApplicationStatus.COMPLETED
                ? finalResult
                : null;
        this.important = important;
        this.manualRegistration = manualRegistration;
        this.memo = memo;
        this.applyDate = applyDate;
        this.interviewDate = interviewDate;
        this.deadlineDate = deadlineDate;
    }

    public void assignApplyEventId(String eventId) {
        this.applyEventId = eventId;
    }

    public void assignInterviewEventId(String eventId) {
        this.interviewEventId = eventId;
    }

    public void assignDeadlineEventId(String eventId) {
        this.deadlineEventId = eventId;
    }

    public void clearApplyEventId() {
        this.applyEventId = null;
    }

    public void clearInterviewEventId() {
        this.interviewEventId = null;
    }

    public void clearDeadlineEventId() {
        this.deadlineEventId = null;
    }

    public void assignNotice(Notice notice) {
        this.notice = notice;
    }
}
